packages/context/src/resolution-session.ts

```diff
@@ -0,0 +1,321 @@
+      () => action(resolutionSession),
+      () => resolutionSession.popBinding(),
+    );
+  }
+
+  /**
+   * Push an injection into the session
+   * @param injection The current injection
+   * @param session The current resolution session
+   */
+  private static enterInjection(
+    injection: Injection,
+    session?: ResolutionSession,
+  ): ResolutionSession {
+    session = session || new ResolutionSession();
+    session.pushInjection(injection);
+    return session;
+  }
+
+  /**
+   * Run the given action with the given injection and session
+   * @param action A function to do some work with the resolution session
+   * @param binding The current injection
+   * @param session The current resolution session
+   */
+  static runWithInjection(
+    action: ResolutionAction,
+    injection: Injection,
+    session?: ResolutionSession,
+  ) {
+    const resolutionSession = ResolutionSession.enterInjection(
+      injection,
+      session,
+    );
+    return tryWithFinally(
+      () => action(resolutionSession),
+      () => resolutionSession.popInjection(),
+    );
+  }
+
+  /**
+   * Describe the injection for debugging purpose
+   * @param injection Injection object
+   */
+  static describeInjection(injection?: Injection) {
+    /* istanbul ignore if */
+    if (injection == null) return injection;
+    const name = getTargetName(
+      injection.target,
+      injection.member,
+      injection.methodDescriptorOrParameterIndex,
+    );
+    return {
+      targetName: name,
+      bindingKey: injection.bindingKey,
+      metadata: injection.metadata,
+    };
+  }
+
+  /**
+   * Push the injection onto the session
+   * @param injection Injection The current injection
+   */
+  pushInjection(injection: Injection) {
+    /* istanbul ignore if */
+    if (debugSession.enabled) {
+      debugSession(
+        'Enter injection:',
+        ResolutionSession.describeInjection(injection),
+      );
+    }
+    this.stack.push({type: 'injection', value: injection});
+    /* istanbul ignore if */
+    if (debugSession.enabled) {
+      debugSession('Resolution path:', this.getResolutionPath());
+    }
+  }
+
+  /**
+   * Pop the last injection
+   */
+  popInjection() {
+    const top = this.stack.pop();
+    assert.equal(
+      top && top.type,
+      'injection',
+      'The top element must be an injection',
+    );
+    const injection = top && (top.value as Injection);
+    /* istanbul ignore if */
+    if (debugSession.enabled) {
+      debugSession(
+        'Exit injection:',
+        ResolutionSession.describeInjection(injection),
+      );
+      debugSession('Resolution path:', this.getResolutionPath() || '<empty>');
+    }
+    return injection;
+  }
+
+  /**
+   * Getter for the current injection
+   */
+  get currentInjection(): Injection | undefined {
+    for (let i = this.stack.length - 1; i >= 0; i--) {
+      if (this.stack[i].type === 'injection')
+        return <Injection>this.stack[i].value;
```

Comment:
The necessity for explicit type cast shows a problem in the design of `ResolutionElement`, because it allows invalid combinations like `{type: 'binding', value: someInjection}`.

Have you considered using a Tagged Union Types to more accurately describe the allowed values and get better type checking from TypeScript?

```ts
export interface BindingResolution {
  type: 'binding'; 
  value: Binding; 
};

export interface InjectionResolution {
  type: 'injection'; 
  value: Injection; 
};

export type ResolutionElement = BindingResolution | InjectionResolution;
```

See https://blog.mariusschulz.com/2016/11/03/typescript-2-0-tagged-union-types and https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions